### PR TITLE
Make scrollbar assertions less aggressive

### LINKED ISSUE #84735: Scrollbar throws due to recent assert changes
## Steps to Reproduce

1. PrimaryScrollController (or any other scroll controller that attaches both to scrollbar and view) -> 
PageTransitionSwitcher ->
SharedAxisTransition ->
Scrollbar ->
View
2. Animate the switcher

Related PR https://github.com/flutter/flutter/pull/81278

<details>
  <summary>Code sample</summary>

```dart
import 'package:flutter/material.dart';
import 'package:animations/animations.dart';

void main() {
  runApp(App());
}

class App extends StatefulWidget {
  @override
  State<App> createState() => _AppState();
}

class _AppState extends State<App> {
  late final views = [
    for (int i = 0; i < 5; i++)
      Scrollbar(
        interactive: true,
        key: ValueKey(i),
        child: ListView.builder(
          itemCount: 100,
          itemBuilder: (context, index) => Text(index.toString()),
        ),
      ),
  ];

  int index = 0;
  void _handleTap() {
    setState(() {
      index += 1;
      if (index == views.length) {
        index = 0;
      }
    });
  }

  final controller = ScrollController();

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'App',
      home: PrimaryScrollController(
        controller: controller,
        child: Scaffold(
          body: Column(
            children: [
              ElevatedButton(onPressed: _handleTap, child: Text('sample text')),
              Expanded(
                child: PageTransitionSwitcher(
                  duration: const Duration(milliseconds: 300),
                  transitionBuilder: (child, animation, secondaryAnimation) {
                    return SharedAxisTransition(
                      transitionType: SharedAxisTransitionType.horizontal,
                      animation: animation,
                      secondaryAnimation: secondaryAnimation,
                      fillColor: Colors.transparent,
                      child: child,
                    );
                  },
                  child: views[index],
                ),
              ),
            ],
          ),
        ),
      ),
    );
  }
}

```
</details>

<details>
  <summary>Logs</summary>

```
════════ Exception caught by scheduler library ═════════════════════════════════
The following assertion was thrown during a scheduler callback:
The PrimaryScrollController is currently attached to more than one ScrollPosition.

The Scrollbar requires a single ScrollPosition in order to be painted.

When the scrollbar is interactive, the associated Scrollable widgets must have unique ScrollControllers. The PrimaryScrollController is used by default for ScrollViews with an Axis.vertical ScrollDirection, unless the ScrollView has been provided its own ScrollController. More than one Scrollable may have tried to use the PrimaryScrollController of the current context.

When the exception was thrown, this was the stack
#0      RawScrollbarState._checkForValidScrollPosition.<anonymous closure>
package:flutter/…/widgets/scrollbar.dart:1101
#1      RawScrollbarState._checkForValidScrollPosition
package:flutter/…/widgets/scrollbar.dart:1125
#2      RawScrollbarState._maybeTriggerScrollbar.<anonymous closure>
package:flutter/…/widgets/scrollbar.dart:1042
#3      SchedulerBinding._invokeFrameCallback
package:flutter/…/scheduler/binding.dart:1143
#4      SchedulerBinding.handleDrawFrame
package:flutter/…/scheduler/binding.dart:1088
#5      SchedulerBinding._handleDrawFrame
package:flutter/…/scheduler/binding.dart:996
#9      _invoke (dart:ui/hooks.dart:163:10)
#10     PlatformDispatcher._drawFrame (dart:ui/platform_dispatcher.dart:259:5)
#11     _drawFrame (dart:ui/hooks.dart:126:31)
(elided 3 frames from dart:async)
════════════════════════════════════════════════════════════════════════════════

```

```
[✓] Flutter (Channel master, 2.3.0-17.0.pre.375, on Microsoft Windows [Version 10.0.19041.1052], locale ru-RU)
    • Flutter version 2.3.0-17.0.pre.375 at C:\dev\src\flutter
    • Upstream repository git@github.com:nt4f04und/flutter.git
    • Framework revision 57a565c706 (19 hours ago), 2021-06-15 18:44:02 -0700
    • Engine revision e0011fa561
    • Dart version 2.14.0 (build 2.14.0-211.0.dev)

[!] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    ✗ cmdline-tools component is missing
      Run `path/to/sdkmanager --install "cmdline-tools;latest"`
      See https://developer.android.com/studio/command-line for more details.
    ✗ Android license status unknown.
      Run `flutter doctor --android-licenses` to accept the SDK licenses.
      See https://flutter.dev/docs/get-started/install/windows#android-setup for more details.

[✓] Chrome - develop for the web
    • Chrome at C:\Program Files\Google\Chrome\Application\chrome.exe

[✓] Visual Studio - develop for Windows (Visual Studio Community 2019 16.7.7)
    • Visual Studio at C:\Program Files (x86)\Microsoft Visual Studio\2019\Community
    • Windows 10 SDK version 10.0.19041.0

[✓] Android Studio (version 4.1.0)
    • Android Studio at C:\Program Files\Android\Android Studio
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b01)

[✓] IntelliJ IDEA Community Edition (version 2020.3)
    • IntelliJ at C:\Program Files\JetBrains\IntelliJ IDEA Community Edition 2020.3.3
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart

[✓] VS Code (version 1.57.0)
    • VS Code at C:\Users\danya\AppData\Local\Programs\Microsoft VS Code
    • Flutter extension version 3.23.0

[✓] Connected device (4 available)
    • sdk gphone x86 (mobile) • emulator-5554 • android-x86    • Android 11 (API 30) (emulator)
    • Windows (desktop)       • windows       • windows-x64    • Microsoft Windows [Version 10.0.19041.1052]
    • Chrome (web)            • chrome        • web-javascript • Google Chrome 91.0.4472.101
    • Edge (web)              • edge          • web-javascript • Microsoft Edge 91.0.864.41

! Doctor found issues in 1 category.
```

</details>

### PR DESCRIPTION
**Pull request by Piinks**
This adjusts the assertions on interactive scrollbars to only be applied when the scrollbar is visible. 

Related issues:
- Fixes https://github.com/flutter/flutter/issues/84735
- https://github.com/flutter/gallery/issues/523

This also adds more documentation and code samples. I cleaned up a bit, removing outstanding todos and fixing some abbreviations in the tests.

For some of the removed todos, I filed issues to actually track the feature requests. 🙃 
- https://github.com/flutter/flutter/issues/84793
- https://github.com/flutter/flutter/issues/84792

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: packages/flutter/lib/src/widgets/scrollbar.dart [hunk 1]
```diff
@@ -665,25 +665,43 @@ class ScrollbarPainter extends ChangeNotifier implements CustomPainter {
...
-/// applied Scrollbars include
+/// ### Interaction
+///
+/// Scrollbars are interactive and can use the [PrimaryScrollController] if
+/// a [controller] is not set. Interactive Scrollbar thumbs can be dragged along
+/// the main axis of the [ScrollView] to change the [ScrollPosition]. Tapping
+/// along the track exclusive of the thumb will trigger a
+/// [ScrollIncrementType.page] based on the relative position to the thumb.
+///
+/// When using the [PrimaryScrollController], it should not be attached to more
```
Comment (dnfield) [inline packages/flutter/lib/src/widgets/scrollbar.dart:685]: s/should/must

File: packages/flutter/lib/src/widgets/scrollbar.dart [hunk 2]
```diff
@@ -665,25 +665,43 @@ class ScrollbarPainter extends ChangeNotifier implements CustomPainter {
...
+/// than one [ScrollPosition]. [ScrollView]s that have not been provided a
+/// [ScrollController] and have a [ScrollView.scrollDirection] of
+/// [Axis.vertical] will automatically attach their ScrollPosition to the
+/// PrimaryScrollController. Provide a unique ScrollController to each
+/// [Scrollable] in this case to prevent having multiple ScrollPositions
+/// attached to the PrimaryScrollController.
```
Comment (dnfield) [inline packages/flutter/lib/src/widgets/scrollbar.dart:691]: An example would really help here.

File: packages/flutter/lib/src/widgets/scrollbar.dart [hunk 3]
```diff
@@ -1212,7 +1286,7 @@ class RawScrollbarState<T extends RawScrollbar> extends State<T> with TickerProv
   @protected
   @mustCallSuper
   void handleThumbPress() {
-    _checkForValidScrollPosition();
+    _debugCheckHasValidScrollPosition();
```
Comment (dnfield) [inline packages/flutter/lib/src/widgets/scrollbar.dart:1289]: Methods called `_debug*` should usually be wrapped in an assert and return a bool.

File: packages/flutter/test/cupertino/scrollbar_test.dart [hunk 4]
```diff
@@ -961,7 +961,13 @@ void main() {
 
     await tester.pumpAndSettle();
 
-    final dynamic exception = tester.takeException();
+    dynamic exception = tester.takeException();
```
Comment (dnfield) [inline packages/flutter/test/cupertino/scrollbar_test.dart:964]: While you're here, can we get rid of `dynamic`?  Perhaps even just changing this to `AssertionError? exception = tester.takeException() as AssertionError?`.

**Comment by davidmartos96** [comment]
I like the proposed changes. Delaying the assertion to when the scrollbar has been interacted with, will help a lot to tell which scrollbar in the tree is incorrect. The error in the gallery project was as a bit confusing.

**Comment by Piinks** [comment]
> Delaying the assertion to when the scrollbar has been interacted with, will help a lot to tell which scrollbar in the tree is incorrect.

This should only delay it until it is visible and available for interaction. :)

**Review by dnfield — APPROVED** [review]
Thanks for the example, really helpful


Lgtm

**Pull request merged**